snapshot/snapshot.go

```diff
@@ -10,96 +10,79 @@ import (
-	if lhs < rhs {
-		return lhs
-	} else {
-		return rhs
-	}
-}
-
-// Random number generator. Implementations should be thread safe.
-type RandomGenerator interface {
-	// @return uint64 a new random number.
-	Random() uint64
-}
-
-// Implementation of RandomGenerator that uses a time seeded random generator.
-type randomGeneratorImpl struct {
-	sync.Mutex
-	random *rand.Rand
-}
-
-func (r *randomGeneratorImpl) Random() uint64 {
-	r.Lock()
-	v := uint64(r.random.Int63())
-	r.Unlock()
-	return v
+type random struct {
+	mu sync.Mutex
+	rr *rand.Rand
 }
 
-var defaultRandomGenerator RandomGenerator = &randomGeneratorImpl{
-	random: rand.New(rand.NewSource(time.Now().UnixNano())),
+func (r *random) Uint64() uint64 {
+	r.mu.Lock()
+	x := r.rr.Int63()
+	r.mu.Unlock()
+	return uint64(x)
 }
 
 // Implementation of Snapshot for the filesystem loader.
 type Snapshot struct {
 	entries map[string]*entry.Entry
+	rand    random
 }
 
 func New() (s *Snapshot) {
-	s = &Snapshot{
+	rr := random{
+		rr: rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	return &Snapshot{
 		entries: make(map[string]*entry.Entry),
+		rand:    rr,
 	}
+}
 
-	return
+func min(lhs, rhs uint64) uint64 {
+	if lhs <= rhs {
+		return lhs
+	}
+	return rhs
 }
 
 func (s *Snapshot) FeatureEnabled(key string, defaultValue uint64) bool {
-	return defaultRandomGenerator.Random()%100 < min(s.GetInteger(key, defaultValue), 100)
+	return s.rand.Uint64()%100 < min(s.GetInteger(key, defaultValue), 100)
 }
 
 // FeatureEnabledForID checks that the crc32 of the id and key's byte value falls within the mod of
 // the 0-100 value for the given feature. Use this method for "sticky" features
 func (s *Snapshot) FeatureEnabledForID(key string, id uint64, defaultPercentage uint32) bool {
-	if e, ok := s.Entries()[key]; ok {
-		if e.Uint64Valid {
-			return enabled(id, uint32(e.Uint64Value), key)
-		}
+	if e := s.entries[key]; e != nil && e.Uint64Valid {
+		return enabled(id, uint32(e.Uint64Value), key)
 	}
-
 	return enabled(id, defaultPercentage, key)
 }
 
 func (s *Snapshot) Get(key string) string {
-	e, ok := s.entries[key]
-	if ok {
+	if e := s.entries[key]; e != nil {
 		return e.StringValue
-	} else {
-		return ""
 	}
+	return ""
 }
 
 func (s *Snapshot) GetInteger(key string, defaultValue uint64) uint64 {
-	e, ok := s.entries[key]
-	if ok && e.Uint64Valid {
+	if e := s.entries[key]; e != nil && e.Uint64Valid {
 		return e.Uint64Value
-	} else {
-		return defaultValue
 	}
+	return defaultValue
 }
 
 // GetModified returns the last modified timestamp for key. If key does not
 // exist, the zero value for time.Time is returned.
 func (s *Snapshot) GetModified(key string) time.Time {
-	if e, ok := s.entries[key]; ok {
+	if e := s.entries[key]; e != nil {
 		return e.Modified
 	}
-
 	return time.Time{}
 }
 
 func (s *Snapshot) Keys() []string {
-	ret := []string{}
+	ret := make([]string, 0, len(s.entries))
 	for key := range s.entries {
 		ret = append(ret, key)
```

Comment:
```suggestion
	ret := make([]string, len(s.entries))
	for i, key := range s.entries {
		ret[i] = key
```

This is faster (?)

Reply:
It would be if `s.entries` we're a slice, but it's a map so this will actually fail to compile.

Reply:
Oops